File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    static int[] dy = {-1, 1, 0, 0};
    static int[] dx = {0, 0, -1, 1};
    static char[] d = {'N', 'S', 'W', 'E'};
    static int[] ret = new int[2];
    static int n, m, sy, sx;
    static char[][] park1;
    static void go(int y, int x, char dir, int size){
        for(int i = 0; i < 4; i++){
            if(d[i] != dir) continue;
            int ny = y + dy[i] * size;
            int nx = x + dx[i] * size;
            // System.out.println(ny + " :: "+ nx);
            if(ny < 0 || nx < 0 || ny >= n || nx >= m) return;
            int k = 0;
            while(k < size){
                k++;
                int yy = y + dy[i]*k;
                int xx = x + dx[i]*k;
                if(yy < 0 || xx < 0 || y >= n || x >= m) return;
                if(park1[yy][xx] == 'X') return;
            }
            ret[0] = ny;
            ret[1] = nx;
        }
        return;
    }
    public int[] solution(String[] park, String[] routes) {
        n = park.length;
        m = park[0].length();
        park1 = new char[n][m];
        for(int i = 0; i < n; i++){
            park1[i] = park[i].toCharArray();
        }
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(park1[i][j] == 'S'){
                    sy = i; sx = j;
                }
            }
        }
        ret[0] = sy;
        ret[1] = sx;
        // for(int i = 0; i < n; i++){
        //     for(int j = 0; j < m; j++){
        //         System.out.print(park1[i][j] + " ");
        //     }
        //     System.out.println();
        // }
        
        for(String s : routes){
            char dir = s.charAt(0);
            int size = s.charAt(2) - '0';
            // System.out.println(dir + " :: "+ size);
            go(ret[0], ret[1], dir, size);
        }
        return ret;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    board = []
    for p in park:
        board.append([c for c in p])
    row, col = len(board), len(board[0])
    ret = []
    for i in range(row):
        for j in range(col):
            if board[i][j] == 'S':
                ret = [i, j]    
    dy,dx = [-1, 0, 1, 0], [0, 1, 0, -1]
    direc = {"N": 0, "E":1, "S":2, "W":3}
    for rou in routes:
        cmd = rou.split()
        size = int(cmd[1])
        dd = direc[cmd[0]]
        ny, nx = ret[0] + dy[dd]*size, ret[1] + dx[dd]*size
        if ny < 0 or ny >= row or nx < 0 or nx >= col:
            continue
        flag = False
        ty, tx = ret
        while tx != nx or ty != ny:
            ty, tx = ty + dy[dd], tx + dx[dd]
            if board[ty][tx] == 'X':
                flag = True
                break
        if flag: continue
        ret = [ny, nx]
        
    return ret
        
        



"""
[조건]
이동시 장애물, 경계밖 확인 => 명령을 무시
모든 명령을 수행한 후 로봇이 놓일 위치 [세로, 가로]
50 * 50 보드

[아이디어]
경계밖 체크 후 장애물 체크 후 좌표 옮기기

[자료구조]
방향벡터, 문자방향인덱스, 결과좌표, 변형보드 

[로직]
1.라우트 순회
2. 각 라우트 별로 다음을 진행 
    2-1. 경계밖인지 체크 => 맞으면 무시
    2-2. 이동시 장애물 있는지 확인 => 맞으면 무시
    2-3. 이동
3. 결과 좌표 반환

"""